test/unit/org/apache/cassandra/cql3/ast/ExpressionEvaluator.java

```diff
@@ -21,126 +21,152 @@
+                    List<Object> accum = new ArrayList<>((List<Object>) lhs);
+                    accum.addAll((List<Object>) rhs);
+                    return accum;
+                }
+                if (lhs instanceof Map)
+                {
+                    Map<Object, Object> accum = new HashMap<>((Map<Object, Object>) lhs);
+                    accum.putAll((Map<Object, Object>) rhs);
+                    return accum;
+                }
                 throw new UnsupportedOperationException("Unexpected type: " + lhs.getClass());
             }
             case SUBTRACT:
             {
-                var lhsOpt = tryEval(e.left);
-                var rhsOpt = tryEval(e.right);
-                if (lhsOpt.isEmpty() || rhsOpt.isEmpty())
-                    return Optional.empty();
-                Object lhs = lhsOpt.get();
-                Object rhs = rhsOpt.get();
                 if (lhs instanceof Byte)
-                    return of((byte) (((Byte) lhs) - ((Byte) rhs)));
+                    return (byte) (((Byte) lhs) - ((Byte) rhs));
                 if (lhs instanceof Short)
-                    return of((short) (((Short) lhs) - ((Short) rhs)));
+                    return (short) (((Short) lhs) - ((Short) rhs));
                 if (lhs instanceof Integer)
-                    return of((int) (((Integer) lhs) - ((Integer) rhs)));
+                    return (int) (((Integer) lhs) - ((Integer) rhs));
                 if (lhs instanceof Long)
-                    return of((long) (((Long) lhs) - ((Long) rhs)));
+                    return (long) (((Long) lhs) - ((Long) rhs));
                 if (lhs instanceof Float)
-                    return of((float) (((Float) lhs) - ((Float) rhs)));
+                    return (float) (((Float) lhs) - ((Float) rhs));
                 if (lhs instanceof Double)
-                    return of((double) (((Double) lhs) - ((Double) rhs)));
+                    return (double) (((Double) lhs) - ((Double) rhs));
                 if (lhs instanceof BigInteger)
-                    return of(((BigInteger) lhs).subtract((BigInteger) rhs));
+                    return ((BigInteger) lhs).subtract((BigInteger) rhs);
                 if (lhs instanceof BigDecimal)
-                    return of(((BigDecimal) lhs).subtract((BigDecimal) rhs));
+                    return ((BigDecimal) lhs).subtract((BigDecimal) rhs);
+                if (lhs instanceof Set)
+                {
+                    Set<Object> accum = new HashSet<>((Set<Object>) lhs);
+                    accum.removeAll((Set<Object>) rhs);
+                    return accum.isEmpty() ? null : accum;
+                }
+                if (lhs instanceof List)
+                {
+                    List<Object> accum = new ArrayList<>((List<Object>) lhs);
+                    accum.removeAll((List<Object>) rhs);
+                    return accum.isEmpty() ? null : accum;
+                }
+                if (lhs instanceof Map)
```

Comment:
this code path is currently untested, only part of this patch that wasn't validated